### ISSUE TITLE
bug(web): autocorrect initial feedback 🚂

### ISSUE THREAD

**Issue by mcdurdin** [keymanapp/keyman#12648]
I've been using Keyman 18 alpha on my phone now for a while, with autocorrect switched on. Here are some initial experience issues:

1. Sometimes space inserts two spaces; I cannot pick exactly what triggers this but once it starts, space will always insert two spaces until I move the insertion point manually
2. Pressing period, comma, etc, does not accept an automatic suggestion -- only spacebar
3. When typing rapidly, one often does not notice if the autocorrect chose wrongly. Backspacing to undo autocorrect only seems to work immediately after the spacebar is pressed; once any other keys are pressed, you have to delete the entire word and retype.
4. Because we don't have learning, wrong autocorrects will always go wrong -- which is super frustrating (e.g. see 23 -> 2nd below). This may be a dealbreaker.

The following may all be related to nrc.en.mtnt. The limitations in the wordlist become more significant when we have autocorrect, e.g.:

1. Numbers get autocorrected, e.g. 23 -> 2nd.
2. We have many words missing in our English dictionary, so rarer words are frequently autocorrected. It will be essential to have 1 and 2 letter words in the dictionary.
3. Curiously we have many words where a longer form is present but a shorter form is not, e.g. we have `authorized` but not `authorize`.
4. Many longer forms are missing, e.g. we have `dance`, `dances`, `dancing`, `dancer`, `dancers`, but not `danced`.
5. It is a bit androgynous when it comes to en-US vs en-UK -- it has some UK forms but not all, e.g. `authorize` but not `authorise`, whereas we have both `behavior` and `behaviour` (and `behavioral` but not `behavioural`). We may need to split en-US and en-UK into separate models for better outcomes.
6. `i` is not corrected to `I`, in English (not sure if one letter words can be resolved by updating the wordlist, or if it will be blocked by lexical model implementation also).

All this said, generally, I find the autocorrect a positive experience. Most aspects work as I would hope and expect, but we need to iron these rough edges out before 18.0 release.

**Comment by jahorton** [comment]
Had an idea over the weekend after looking at this - went ahead and threw together #12877 to address (and generalize) the "numbers" problem.

**Comment by jahorton** [comment]
> Pressing period, comma, etc, does not accept an automatic suggestion -- only spacebar

Refer to #12013 - I believe the same ideas for punctuation handling there would also pay off here; I believe the desired logic to be very similar for both cases.

**Comment by jahorton** [comment]
> When typing rapidly, one often does not notice if the autocorrect chose wrongly. Backspacing to undo autocorrect only seems to work immediately after the spacebar is pressed; once any other keys are pressed, you have to delete the entire word and retype.

For this one... I think the "context tracker" system could, in theory, be leveraged to detect this case with some additional work.  We're basically looking to revert the suggestion after a delay in this case, rather than immediately after its application.  This is also presently a limitation outside of autocorrect, so it's something I could look into for the current release cycle if desired.

**Comment by jahorton** [comment]
> 1. Sometimes space inserts two spaces; I cannot pick exactly what triggers this but once it starts, space will always insert two spaces until I move the insertion point manually

Extra detail:  regardless of whether or not any suggestion-acceptance is involved.  It sounds to me like it may be related to the space-based suggestion acceptance, but with bookkeeping gone wrong.  Just my gut instinct speaking here at present, though.

**Comment by jahorton** [comment]
> 4\. Because we don't have learning, wrong autocorrects will always go wrong -- which is super frustrating (e.g. see 23 -> 2nd below). This may be a dealbreaker.

Note that this may prove difficult on iOS; the app extension cannot write to the App Group in order to share data with the app.  This means we lack a straightforward way to allow users to clearly manage or manipulate any learned data.

**Comment by mcdurdin** [comment]
> This means we lack a straightforward way to allow users to clearly manage or manipulate any learned data for any users who leave it off or disabled.

This is true for any keyboard app on iOS. I don't think it is a dealbreaker -- the UX for managing learned data would need to let users know that they have to toggle that setting to manage. Or we could even add some UX in the system keyboard app extension itself.

**Comment by jahorton** [comment]
> * Curiously we have many words where a longer form is present but a shorter form is not, e.g. we have `authorized` but not `authorize`.
> 
> * Many longer forms are missing, e.g. we have `dance`, `dances`, `dancing`, `dancer`, `dancers`, but not `danced`.

I believe these cases are best handled by #3058 - morphological models would be well-suited to modeling part-of-speech permutations and determining some form of word-frequency based on each word's constituent affixes.  We'd also likely want similar handling for "extending" previous suggestions with new suffixes (dance - > dancing, etc).

**Comment by mcdurdin** [comment]
Perhaps, but developing a morphological model is going take a very long time -- both for the Keyman side of things, and then for the English model; we could, quite a lot more quickly, refresh the wordlist with some analysis (or there may be other wordlists out there with a better distribution than the one we already have?) Depends a bit on model-encoding because we shortened the English wordlist for performance as well.

**Comment by jahorton** [comment]
> we could, quite a lot more quickly, refresh the wordlist with some analysis

Are we willing to put in similar efforts for _each_ model?  I suspect that many of our wordlist models will have a similar problem regarding the words of the languages they model.  We'd probably need the community to do the refresh for those once we're better able to support a higher wordlist count via model-encoding.

**Comment by mcdurdin** [comment]
But who is going to put in the effort to build morphological models for each language? That's even harder than building a full wordlist and takes specialist skills, unlike wordlist collection which is fairly straightforward

**Comment by jahorton** [comment]
> Perhaps, but developing a morphological model is going take a very long time -- both for the Keyman side of things, and then for the English model; we could, quite a lot more quickly, refresh the wordlist with some analysis (or there may be other wordlists out there with a better distribution than the one we already have?) Depends a bit on model-encoding because we shortened the English wordlist for performance as well.

So, should we make some sort of issue to this effect?  And is there anything else from this PR not yet reflected in its own separate issue?  It'd be nice to close this one off in favor of specialized issues.

**Comment by mcdurdin** [comment]
> So, should we make some sort of issue to this effect?

For now, tracking against nrc.en.mtnt is sufficient; see below.

> And is there anything else from this PR not yet reflected in its own separate issue?

Sometimes space inserts two spaces; I cannot pick exactly what triggers this but once it starts, space will always insert two spaces until I move the insertion point manually.

And conversely, sometimes selecting a word from the banner does not insert a space afterwards. I have not been able to identify a trigger for this.

--> tracking these space-related issues in #16400.

---

I have added the nrc.en.mtnt-specific items to keymanapp/lexical-models#242, including the suggestion to add more forms of words.

I don't know if the `i` -> `I` change can be done in nrc.en.mtnt (e.g. by adding `I` with a very high frequency?) or if that will be blocked? See also #8836?

---

So yes, I think we can close this.

**Comment by mcdurdin** [comment]
Remaining issues split out for follow-up.
